### ISSUE TITLE
Decouple test definition from test running in verification.py

### ISSUE THREAD

**Issue by jeromekelleher** [tskit-dev/msprime#1056]
#1035 made a good start on refactoring the verification.py script, which is a big horrible mess.

The next step in the process of making this comprehensible and easier to work with is to decouple the test *definitions* from the actual running of the tests.

What I'd like to see is something like the following 
```python
@attr.s
class Test:
    """
    The superclass of all tests. Each test must define a name, a group
    and a run method.
    """
    name = attr.ib(type=str)
    group = attr.ib(type=str)

    def run(self, output_dir):
        raise NotImplementedError()


@attr.s
class MsCommandLine(Test):

    command = attr.ib(type=str)

    def run(self, output_dir):
        # Rather than printing like we did before, let's use the logging infrastructure. 
        logging.info("running", self.command)
        # Port all the run stuff in here.


def register_ms_command_line_tests(runner):
    def register(name, command):
        runner.register(MsCommandLine(name=name, group="ms", command=command))

    register("size-change1", "10 10000 -t 2.0 -eN 0.1 2.0")
    register("growth-rate-change1", "10 10000 -t 2.0 -eG 0.1 5.0")
    register("growth-rate-2-pops1", "10 10000 -t 2.0 -I 2 5 5 2.5 -G 5.0")
    # etc.

@attr.s
class TestRunner:
    """
    Class responsible for registering all known tests and running
    them.
    """
    tests = attr.ib(factory=list)
    groups = attr.ib(factory=set)

    def register(self, test):
        self.tests.append(test)
        self.groups.add(test.group)

    def run(self, output_dir, names=None, group=None):
        tests = self.tests
        if names is not None:
            tests = [test for test in self.tests if test.name in names]
        elif group is not None:
            tests = [test for test in self.tests if test.group == group]
        for test in tests:
            test.run(output_dir)
```

Thus, the TestRunner does nothing except register and run the tests. All the actual test behaviour is encapsulated in the classes. The ``group`` concept is flexible, and doesn't have to be tied to a particular test type or class.

Some of the tests are most bespoke where there's very little sharing of code between different tests within a group. The simplest thing for now there may be to make a different class for each test. Hopefully we'll be able to factor out some of the common code in those later.

How does this look @saunack?

**Comment by saunack** [comment]
Looks good. I got most of it. I'll get back to you in case of difficulties

**Comment by saunack** [comment]
@jeromekelleher  I'm having some trouble understanding what exactly will be in the `run()` function, since `add_ms_instance()` uses several functions of the `MsTest` class. Should I pull out all the code from `add_ms_instance()` and use it in `run()`? But that would be extracting out the test definitions as well as code from their original classes. Can you clarify this a bit please?

**Comment by saunack** [comment]
Also, another thing: some groups like `dtwf` have multiple tests like `add_dtwf_vs_coalescent_2_pops_massmigration()`, `add_dtwf_vs_coalescent_2_pop_growth()` etc. unlike `ms` in which all tests share the same skeleton `add_ms_instance()`. I'm confirming if we want to create a different class and registering functions for each.

**Comment by jeromekelleher** [comment]
> @jeromekelleher I'm having some trouble understanding what exactly will be in the run() function, since add_ms_instance() uses several functions of the MsTest class. Should I pull out all the code from add_ms_instance() and use it in run()? But that would be extracting out the test definitions as well as code from their original classes. Can you clarify this a bit please?

Good question @saunack! The way I'm imagining this is that we'll basically transfer the class hierarchy you currently have to Tests. Each test knows how to ``run`` itself, and we should be able to transfer the code across between the classes. For example, the MsCmdTest can look something like this
```python
class MsCmdTest(Test):
    # stuff
   def run(self, output_dir):
       self._run_coalescent_stats(output_dir)
       self._run_mutation_stats(output_dir)
   
   def _run_coalescent_stats(self, output_dir):
       # Code ported from current defs 

    # etc
```

**Comment by jeromekelleher** [comment]
> Also, another thing: some groups like dtwf have multiple tests like add_dtwf_vs_coalescent_2_pops_massmigration(), add_dtwf_vs_coalescent_2_pop_growth() etc. unlike ms in which all tests share the same skeleton add_ms_instance(). I'm confirming if we want to create a different class and registering functions for each.

Anywhere that we have a unique function to run the test, we'll need a new class. We can't do the tricks with defining local functions, because we need to be able to pickle the tests when we'll be parallelising this.  This is a bit tedious, but I hope that once we've rearranged things, it'll become a little clearer how we can factor out common code from the tests into superclasses.

So, in your example above, we'd have something like
```python

@attr.s
class DtwfVsCoalescentTest(Test):
    simulate_args=attri.ib(factory=dict)
    
    def run(self, output_dir):
          self._run_dtwf_coalescent_comparison( 
                output_dir,
                self.name, 
                 **self.simulate_args)

     # Copy the definition of _run_dtwf_coalescent_comparison here
     def _run_dtwf_coalescent_comparison()
        # etc


def  dtwf_vs_coalescent_2_pops_massmigration():
        population_configurations = [
            msprime.PopulationConfiguration(sample_size=10, initial_size=1000),
            msprime.PopulationConfiguration(sample_size=10, initial_size=1000),
        ]
        recombination_map = msprime.RecombinationMap([0, int(1e6)], [1e-8, 0])
        demographic_events = [
            msprime.MassMigration(time=300, source=1, destination=0, proportion=1.0)
        ]
        return DtwfVsCoalescentTest(
            name = "dtwf_vs_coalescent_2_pops_massmigration",
            simulate_args=dict(
                population_configurations=population_configurations,
                demographic_events=demographic_events,
                num_replicates=300,
                recombination_map=recombination_map,
            )


def register_dtwf_vs_coalescent_tests(runner):
    tests = [
           dtwf_vs_coalescent_2_pops_massmigration(),
           # all the other tests
    ]
    for test in tests:
          tests.group = "dtwf_vs_coalescent"
          runner.register(test)
```

See what I mean?

**Comment by jeromekelleher** [comment]
FYI @saunack, we merged an extra set of tests yesterday in #1032. This added two extra test classes, and it should merge cleanly (hopefully) onto any changes you're making. It would be a good idea to bring your code up to date now though, by rebasing off upstream/master.

**Comment by jeromekelleher** [comment]
The new classes will essentially be the same as the classes that you currently have @saunack, so it's really a fairly minor refactoring from there. The main idea is that we have to change how the tests are registered and run. Instread of a big a class that does essentially everything, we have bunch of test *instances*. These are then much more flexible, and will lend themselves directly to being run in parallel. The current approach of using the ``self._instances`` dictionary won't.

**Comment by jeromekelleher** [comment]
closed in #1068
